sklearn/datasets/_svmlight_format_fast.pyx

```diff
@@ -113,3 +113,188 @@ def _load_svmlight_file(f, dtype, bint multilabel, bint zero_based,
+
+        if multilabel:
+            labels_str = ""
+            first = True
+            for j in range(num_labels):
+                val = y[i,j]
+                if val != 0:
+                    if not first:
+                        labels_str += ","
+                    first = False
+                    labels_str += label_pattern % j
+        else:
+            labels_str = label_pattern % y[i,0]
+
+        if query_id is not None:
+            feat = (labels_str, query_id[i], s)
+        else:
+            feat = (labels_str, s)
+
+        f.write((line_pattern % feat).encode("ascii"))
+
+cdef inline str get_row_string(int_or_float1[:,:] X, Py_ssize_t row, str value_pattern, bint one_based):
+    cdef:
+        Py_ssize_t row_length = X.shape[1]
+        array.array[long long] int_template = array.array('q',[])
+        array.array[cython.double] float_template = array.array('d',[])
+        array.array x_inds
+        array.array x_vals
+        Py_ssize_t x_nz_used = 0
+
+    if int_or_float1 in cython.floating:
+        x_vals = array.clone(float_template, row_length, False)
+    else:
+        x_vals = array.clone(int_template, row_length, False)
+    x_inds = array.clone(int_template, row_length, False)
+
+    for j in range(row_length):
+        val = X[row,j]
+        if val!=0:
+            x_inds[x_nz_used]=j
+            x_vals[x_nz_used]=val
+            x_nz_used += 1
+    return " ".join(value_pattern % (j+one_based, val) for i, (j, val) in enumerate(zip(x_inds, x_vals)) if i < x_nz_used)
+
+def _dump_svmlight_file_general(X, y, f, bint multilabel, bint one_based, int_or_longlong[:] query_id, bint X_is_sp, bint y_is_sp):
+    cdef bint X_is_integral = X.dtype.kind == "i"
+    if X_is_integral:
+        value_pattern = "%d:%d"
+    else:
+        value_pattern = "%d:%.16g"
+    if y.dtype.kind == "i":
+        label_pattern = "%d"
+    else:
+        label_pattern = "%.16g"
+
+    line_pattern = "%s"
+    if query_id is not None:
+        line_pattern += " qid:%d"
+    line_pattern += " %s\n"
+
+    cdef:
+        signed long long[:,:] X_dense_int
+        double[:,:] X_dense_float
+        Py_ssize_t num_labels = y.shape[1]
+        Py_ssize_t x_len = X.shape[0]
+        Py_ssize_t row_length = X.shape[1]
+        Py_ssize_t i
+        Py_ssize_t j
+        Py_ssize_t col_start
+        Py_ssize_t col_end
+        bint first
+        array.array[long long] int_template = array.array('q',[])
+        array.array[cython.double] float_template = array.array('d',[])
+        Py_ssize_t x_nz_used
+        array.array x_inds
+        array.array x_vals
+    for i in range(x_len):
+        x_nz_used = 0
+
+        if X_is_sp:
+            span = slice(X.indptr[i], X.indptr[i + 1])
+            row = zip(X.indices[span], X.data[span])
+            s = " ".join(value_pattern % (j + one_based, x) for j, x in row)
+        else:
+            #s = get_row_string(X, i, value_pattern, one_based)
```

Comment:
Can we get away with:

```python
def get_row_string(int_or_float1[:,:] X, ...):
```

so we can call:

```python
s = get_row_string(X, i, value_pattern, one_based)
```

here? As long as the performance is better than `main`, I am okay with a little more Python interaction here.